applications/ThermalDEMApplication/custom_utilities/heat_map_utilities.cpp

```diff
@@ -0,0 +1,206 @@
+//  Kratos Multi-Physics - ThermalDEM Application
+//
+//  License:       BSD License
+//                 Kratos default license: kratos/license.txt
+//
+//  Main authors:  Rafael Rangel (rrangel@cimne.upc.edu)
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "heat_map_utilities.h"
+
+namespace Kratos {
+  //-----------------------------------------------------------------------------------------------------------------------
+  HeatMapUtilities::HeatMapUtilities() {}
+  HeatMapUtilities::~HeatMapUtilities() {}
+
+  //-----------------------------------------------------------------------------------------------------------------------
+  void HeatMapUtilities::ExecuteInitialize(ModelPart& rModelPart) {
+    const ProcessInfo& r_process_info = rModelPart.GetProcessInfo();
+    if (!r_process_info[HEAT_MAP_GENERATION_OPTION])
+      return;
+
+    // Store properties
+    array_1d<int, 3> subdivisions = r_process_info[HEAT_MAP_SUBDIVISIONS];
+    mDimX = subdivisions[0];
+    mDimY = subdivisions[1];
+    mDimZ = subdivisions[2];
+
+    // Initialize global maps with zeros
+    ResetMap(mGlobalHeatMapGenerationDampingPP);
+    ResetMap(mGlobalHeatMapGenerationDampingPW);
+    ResetMap(mGlobalHeatMapGenerationSlidingPP);
+    ResetMap(mGlobalHeatMapGenerationSlidingPW);
+    ResetMap(mGlobalHeatMapGenerationRollingPP);
+    ResetMap(mGlobalHeatMapGenerationRollingPW);
+
+    // Initialize local maps with zeros
+    const int num_of_particles = rModelPart.NumberOfElements();
+    #pragma omp parallel for schedule(dynamic, 100)
+    for (int i = 0; i < num_of_particles; i++) {
+      ModelPart::ElementsContainerType::iterator it = rModelPart.GetCommunicator().LocalMesh().Elements().ptr_begin() + i;
```

Comment:
you can create the ptr_begin outside the loop for efficiency and then always add to the same